docs: update state management documentation (#781)
widgetti/solara: solara/website/pages/documentation/getting_started/content/05-fundamentals/50-state-management.md

In this case, the `SomeAppSpecificComponent` is not reusable in the sense that a second component has a different state. The `color` variable is global and shared across all components. This component is meant to be used only once, and mainly helps to organize the code.

You may have heard that globals are considered a bad practice. As with many things, it depends on the context. A possible downside of using a global is that it does not
allow you to create multiple instances of the same component with different states. But if the state reflects application state, there is by definition only one instance
of it needed.

### Local component state using solara.use_reactive

> If you do need state that is specific to a component, you should use [`solara.use_reactive`](/documentation/api/hooks/use_reactive) hook. These hooks allow you to create local state variables that are scoped to a specific component. This approach is useful when you want to encapsulate state within a component, making it self-contained and modular. Local state management is suitable for situations where state changes only affect the component and do not need to be shared across the application.

Reviewer: ```suggestion
If you do need state that is specific to a component, you should use [`solara.use_reactive`](/documentation/api/hooks/use_reactive) hook. This hook allows you to create local state variables that are scoped to a specific component. This approach is useful when you want to encapsulate state within a component, making it self-contained and modular. Local state management is suitable for situations where state changes only affect the component and do not need to be shared across the application.
```

```

### Local component state using solara.use_state (not recommended)

[`solara.use_state`](/documentation/api/hooks/use_state) is a hook that might be a bit more familiar to React developers. It also allows you to create local state variables that are scoped to a specific component, however, instead of using reactive variables, it uses a tuple of a value and a setter function.

We generally recommend using `use_reactive` over `use_state` as it is more easy to refactor between global application state and local component state by switching between `use_reactive` and `reactive`. There is no equivalent for `use_state` at the global level.

> If we take the previous example and replace `use_reactive` by `use_reactive`, we get:

Reviewer: ```suggestion
If we take the previous example and replace `use_reactive` by `use_state`, we get:
```

```

## Mutation pittfalls

In Python, strings, numbers, and tuples are immutable. This means that you cannot change the value of a string, number, or tuple in place.
Instead, you need to create a new object and assign that to a variable.

```python
a = 1
b = a
# a.value = 2  # ERROR: numbers are immutable
a = 2  # instead, re-assign a new value, the number 1 will not change
assert b == 1 # b is still 1
```

> However, many objects in Python are mutable, including lists and dictionaries or user defined classes. This means that you can change the value of a list, dictionary, or user defined class in place without creating a new object.

Reviewer: ```suggestion
However, many objects in Python are mutable, including lists, dictionaries, and potentially user defined classes. This means that you can change the value of a list, dictionary, or user defined class in place without creating a new object.
```

```

Although Solara could potentially track mutations of lists and dictionaries, that would be difficult to do for user defined classes, since we would need to know which methods mutate the object and which do not. Therefore, we have chosen not to include
any magic tracking of mutations in Solara, and instead require you to re-assign a new value to a reactive variable if you want to trigger a re-render.

```python hl_lines="5"
import solara

reactive_list = solara.reactive([1, 2, 3])
# Instead, re-assign a new value
reactive_list.value = [*reactive_list.value, 4]  # GOOD: re-assign a new list
```

### Not mutating dictionaries

A similar pattern applies to dictionaries.

```python
import solara

reactive_dict = solara.reactive({"a": 1, "b": 2})
reactive_dict.value = {**reactive_dict.value, "c": 3}  # GOOD: re-assign a new dictionary
# deleting a key
reactive_dict.value = {k: v for k, v in reactive_dict.value.items() if k != "a"}  # GOOD: re-assign a new dictionary
# deleting a key (method 2)
dict_copy = reactive_dict.value.copy()
del dict_copy["b"]
reactive_dict.value = dict_copy  # GOOD: re-assign a new dictionary
```

### Not mutating user defined classes

> Or user defined classes, like a Panda DataFrame.

Reviewer 2: ```suggestion
Or user defined classes, like a Pandas DataFrame.
```

Author: Thanks!

```python
import solara
import pandas as pd

reactive_df = solara.reactive(pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]}))
# reactive_df.value.append({"a": 4, "b": 7})  # BAD: mutating a DataFrame is not observable in Python
df_copy = reactive_df.value.copy(deep=True)  # for Pandas 3, deep=True is not necessary
df_copy = df_copy.append({"a": 4, "b": 7}, ignore_index=True)
reactive_df.value = df_copy  # GOOD: re-assign a new DataFrame
```

## Creating a store

> Using reactive variables is a powerful way to manage state in your Solara applications. However, as your application grows, you may find that you need a more structured way to manage your state..

Reviewer: ```suggestion
Using reactive variables is a powerful way to manage state in your Solara applications. However, as your application grows, you may find that you need a more structured way to manage your state.
```
